### PR TITLE
Hashpin sensitive dependencies and configure dependabot to update them automatically

### LINKED ISSUE #75620: CI: Hashpin sensitive workflow dependencies and install dependabot to update them
Hi, I'm Diogo and I'm back (see #69736). At the issue #69736 I suggested you to set minimal permissions to your workflows. Now I'm coming back to suggest a modification that would provide extra safety for the workflows that yet require dangerous permissions (e.g., contents: write).

### Problem

Your workflows [release-binaries.yml](https://github.com/llvm/llvm-project/blob/main/.github/workflows/release-binaries.yml) and [release-tasks.yml](https://github.com/llvm/llvm-project/blob/main/.github/workflows/release-tasks.yml) are using dangerous permissions while running external dependencies pinned only by tag. Both are using external unpinned GitHub Actions, and the release-tasks.yml is also installing unhashed Python dependencies (see https://pip.pypa.io/en/stable/topics/secure-installs/#hash-checking-mode). Those patterns could be dangerous because if any of those actions get hijacked (and at the end they're all repositories and are susceptible to attacks like any other), an attacker could change the code that your tags point to, gaining access to your secrets and/or write permissions to your repository.

### Proposed Solution

A solution for this problem would be to hash-pin those sensitive dependencies. For the GitHub Actions, we'd point the actions to the very specific commit of that release. It follows an example of the change:

```
- uses: r-lib/actions/pr-fetch@v1 
would become
- uses: r-lib/actions/pr-fetch@11a22a908006c25fe054c4ef0ac0436b1de3edbe # v1.3.1
```

And this would enforce that your action is always running at the expected code.

For the python dependency, we could install them using Python [hash-checking mode](https://pip.pypa.io/en/stable/topics/secure-installs/#hash-checking-mode).

The only downsize of those solutions is that it gets trickier to manually update the version of the actions as they get out-of-date, but that can be solved by using a Dependency-Update-Tool (like dependabot or renovatebot). We can configure it to update all workflow dependencies in a single monthly PR, for example -- for security updates, the tool would immediately create the PRs regardless of the frequency set. For the case of the hash-pinning, the PRs would still keep a comment with the human-readable version used =).

### Conclusion

I'll take the liberty of raising a PR implementing my suggestions, so that it becomes easier for you to evaluate. Let me know if you have any questions or concerns.

### PR DESCRIPTION
**Pull request by diogoteles08**
Closes #75620 

As I mentioned on the issue, this PR aims to hash-pin the CI dependencies used on sensitive context -- i.e., they either are called with write permissions, or are being used to build critical artifacts like a release. In summary, this PR brings 3 changes:

1. Hash pin GitHub Actions called on sensitive context
2. Hash pin python dependencies used on sensitive context
3. Configure dependabot to automatically update those hashes
I'm further explaining the steps bellow.

The dependencies in format of GitHub Actions, I simply hash-pinned them. I also made sure to keep the human-readable version as comments at the same line.

At the [release-tasks.yml](https://github.com/llvm/llvm-project/blob/main/.github/workflows/release-tasks.yml) file, I've changed the installation method of some python dependencies to install them considering their hashpinning. That required the generation of a requirements file that had all the correct hashes, and for that I used [pip-tools](https://pypi.org/project/pip-tools/2.0.0/).

While configuring dependabot, I set it to send a monthly PR updating all the GitHub Actions, and a weekly PR to update any python dependency required by [/llvm/docs/requirements.txt](https://github.com/llvm/llvm-project/blob/main/llvm/docs/requirements.txt).

Let me know if you have any questions or concerns, I'd be happy to clarify and help.

Thanks!

### CODE REVIEW

**Review by tstellar — COMMENTED** [review]
We have several of these requirements.txt files scattered through out the repo, should we make this change with all of them?

**Comment by diogoteles08** [comment]
> We have several of these requirements.txt files scattered through out the repo, should we make this change with all of them?

Hi @tstellar! I'd say it depends on how the dependencies on the requirements.txt are used. It's recommended to use hash-checking installs for production run-time dependencies or dependencies used to build critical artifacts such as releases. But it might be debatable if it's worth it to hash-check every install of test or development dependencies.

The main point for this uncertainty is that, although hash-checking the installations are most secure, it's a bit more difficult to maintain, and it might also be useful to have some dependencies unpinned -- specially test dependencies, that could provide a wider range of versions tested. For the maintenance, Dependabot should be enough to keep your current dependencies updated and you shouldn't need to manually look for their hashes, but in case you want to start using a new dependency, you'd need to re-run the [pip-tools](https://pypi.org/project/pip-tools/2.0.0/) to generate the hashes for the new dependency.

**Comment by tru** [comment]
I think this looks good. But we have (as Tom said) some other python dependencies that we probably should hash as well. The code formatter workflow and github pr workflow comes to mind. I am happy to handle these in other PR's so we can merge this one.

**Review by tru — APPROVED** [review]

**Comment by diogoteles08** [comment]
> I think this looks good. But we have (as Tom said) some other python dependencies that we probably should hash as well. The code formatter workflow and github pr workflow comes to mind. I am happy to handle these in other PR's so we can merge this one.

Sounds great to me! Let me know if you need any help on the other PRs, I'd be happy to help

File: .github/dependabot.yml
Comment (boomanaiden154) [inline .github/dependabot.yml:14]: Maybe change this to at least monthly too? Weekly I think would create a lot of unnecessary churn.

**Review by boomanaiden154 — COMMENTED** [review]
No major issues on my end. Others have already pointed out other scripts/CI jobs using different `requirements.txt` files.

Most of my experience with dependabot is it being slightly worse than useless, but we can try it out for a bit and see how it does with these scripts. It would probably be a lot easier if we had good integration/unit testing for them, but that's hard to do with CI.

File: .github/dependabot.yml
Comment (tstellar) [reply .github/dependabot.yml:14]: I agree with @boomanaiden154 I think starting with monthly would be best.

Comment (diogoteles08) [reply .github/dependabot.yml:14]: Great! Sorry for the late reply, I'll update the PR right away

**Review by tru — APPROVED** [review]

**Review by tstellar — APPROVED** [review]

**Comment by tru** [comment]
@diogoteles08 do you need us to merge this or do you have write acccess?

**Comment by boomanaiden154** [comment]
> @diogoteles08 do you need us to merge this or do you have write acccess?

He's not part of the LLVM organization and thus doesn't have write access to the monorepo, so someone else will have to merge it.

**Comment by diogoteles08** [comment]
> He's not part of the LLVM organization and thus doesn't have write access to the monorepo, so someone else will have to merge it.

That's right, I can't merge it ☹️ 
I just updated the branch to be up to date with last main changes. @tru, can you merge it?

**Pull request merged**
